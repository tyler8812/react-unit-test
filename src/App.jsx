import "./App.css";
import { useState, useEffect } from "react";
import React from "react";
import { formatUserName } from "./utils";
import axios from "axios";

function App() {
  const [users, setUsers] = useState([]);
  useEffect(() => {
    let mounted = true;
    const getUsers = async () => {
      const response = await axios.get(
        "https://jsonplaceholder.typicode.com/users"
      );
      if (mounted) {
        setUsers(response.data);
      }
    };
    getUsers();
  }, []);
  return (
    <div className="App">
      <div>Users:</div>
      {users.length === 0 ? (
        <div key="loading">Loading users...</div>
      ) : (
        <ul data-testid="user-list">
          {users.map((user) => (
            <li key={user.name + user.username} data-testid="user-item">
              {user.name + formatUserName(user.username)}
            </li>
          ))}
        </ul>
      )}
    </div>
  );
}

export default App;
