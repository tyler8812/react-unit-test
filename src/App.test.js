import {
  render,
  screen,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import App from "./App";
import axios from "axios";

jest.mock("axios");

const fakeUsers = [
  {
    id: 1,
    name: "Test User 1",
    username: "testuser1",
  },
  {
    id: 2,
    name: "Test User 2",
    username: "testuser2",
  },
];

describe("App component", () => {
  // test with api
  // test("it renders", () => {
  //   render(<App />);
  //   expect(screen.getByText("Users:")).toBeInTheDocument();
  // });
  // test("it displays a list of users", async () => {
  //   render(<App />);
  //   expect(screen.getByTestId("user-list")).toBeInTheDocument();
  // });
  // test("it displays a list of users", async () => {
  //   render(<App />);
  //   const userList = await waitFor(() => screen.getByTestId("user-list"));
  //   expect(userList).toBeInTheDocument();
  // });
  // mock (fake data not using api)
  // test("it renders", async () => {
  //   axios.get.mockResolvedValue({ data: fakeUsers });
  //   render(<App />);
  //   expect(await screen.findByText("Users:")).toBeInTheDocument();
  // });
  // test("it displays a list of users", async () => {
  //   axios.get.mockResolvedValue({ data: fakeUsers });
  //   render(<App />);
  //   const userList = await waitFor(() => screen.getByTestId("user-list"));
  //   expect(userList).toBeInTheDocument();
  // });
  // test("it displays a row for each user", async () => {
  //   axios.get.mockResolvedValue({ data: fakeUsers });
  //   render(<App />);
  //   const userList = await waitFor(() => screen.findAllByTestId("user-item"));
  //   expect(userList).toHaveLength(2);
  // });

  // snapshot test
  // test("it renders a correct snapshot", async () => {
  //   axios.get.mockResolvedValue({ data: fakeUsers });
  //   const { asFragment } = render(<App />);

  //   expect(asFragment()).toMatchSnapshot();
  // });

  test("it renders a correct snapshot", async () => {
    axios.get.mockResolvedValue({ data: fakeUsers });
    const { asFragment } = render(<App />);
    await waitForElementToBeRemoved(screen.getByText("Loading users..."));
    expect(asFragment()).toMatchSnapshot();
  });
});
