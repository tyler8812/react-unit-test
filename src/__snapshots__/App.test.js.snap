// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component it renders a correct snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <div>
      Users:
    </div>
    <ul
      data-testid="user-list"
    >
      <li
        data-testid="user-item"
      >
        Test User 1@testuser1
      </li>
      <li
        data-testid="user-item"
      >
        Test User 2@testuser2
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
